assertion failure in `identifierLocFromIndex`

### Zig Version

0.13

### ZLS Version

0.14.0-dev.147+dd78968

### Client / Code Editor / Extensions

_No response_

### Steps to Reproduce and Observed Behavior

ive been trying to see if i can use new ZLS on projects using older Zig. it usually does seem to work, but I sometimes get a crash after long usage. using a debug build is not too bad (sometimes get lag spikes), so if theres anything i can do on my install to log debug information, i would love to.

going by the stack trace, it's in `textDocument/documentHighlight`. i don't know enough zls/lsp internals to do much with this info.

### Expected Behavior

no assertion failure

### Relevant log output

```shell
info  ( main ): Starting ZLS      0.14.0-dev.147+dd78968 @ '/Users/dave/code/zls/zig-out/bin/zls'
info  ( main ): Message Tracing:  false
info  ( main ): Log Level:        debug
info  ( main ): Log File:         /Users/dave/Library/Caches/zls/zls.log
info  (server): Client Info:      Visual Studio Code-1.93.0
debug (server): Offset Encoding:  utf-16
info  (server): Workspace Folder: 'file:///Users/dave/code/bun'
info  (server): Set config option 'builtin_path' to "/Users/dave/Library/Caches/zls/builtin.zig"
info  (server): Set config option 'zig_lib_path' to "/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib"
info  (server): Set config option 'zig_exe_path' to "/Users/dave/.zig/bin/zig"
info  (server): Set config option 'build_runner_path' to "/Users/dave/Library/Caches/zls/build_runner/b0f5526cec8ba7d4a2bf68536fe92d28/build_runner.zig"
info  (server): Set config option 'global_cache_path' to "/Users/dave/Library/Caches/zls"
debug (server): Took 41ms to process request-0-initialize on Thread 2620455
debug (server): Dynamically registering method 'workspace/didChangeConfiguration'
debug (server): Took 0ms to process notification-initialized on Thread 2620455
info  (store ): Loaded build file 'file:///Users/dave/code/bun/build.zig'
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/import_record.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/import_record.zig'
debug (server): Took 1ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/bun.js/api/JSBundler.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/bun.js/api/JSBundler.zig'
debug (server): Took 3ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/cli/build_command.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/cli/build_command.zig'
debug (server): Took 4ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/string_immutable.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/string_immutable.zig'
debug (server): Took 10ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/kit/DevServer.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/kit/DevServer.zig'
debug (server): Took 2ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/kit/kit.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/kit/kit.zig'
debug (server): Took 0ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Opened document 'file:///Users/dave/code/bun/vendor/zig/lib/std/mem.zig'
debug (server): Took 9ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/bundler/bundle_v2.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/bundler/bundle_v2.zig'
debug (server): Took 21ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/crash_handler.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/crash_handler.zig'
debug (server): Took 3ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/js_parser.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/js_parser.zig'
debug (server): Took 35ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/js_ast.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/js_ast.zig'
debug (server): Took 13ms to process notification-textDocument/didOpen on Thread 2620455
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/bundler.zig' as 'file:///Users/dave/code/bun/build.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/src/bundler.zig'
debug (server): Took 5ms to process notification-textDocument/didOpen on Thread 2620455
debug (server): Took 0ms to process response-"register-workspace/didChangeConfiguration" on Thread 2620455
info  (server): Set config option 'enable_build_on_save' to false
info  (server): Set config option 'enable_autofix' to true
info  (server): Set config option 'semantic_tokens' to "partial"
info  (server): Set config option 'zig_lib_path' to "/Users/dave/code/bun/vendor/zig/lib"
info  (server): Set config option 'zig_exe_path' to "/Users/dave/code/bun/.cache/zig/zig.exe"
debug (server): Took 43ms to process response-"i_haz_configuration" on Thread 2620455
error (store ): Failed to execute build runner to collect build configuration, command:
/Users/dave/.zig/bin/zig build --build-runner /Users/dave/Library/Caches/zls/build_runner/b0f5526cec8ba7d4a2bf68536fe92d28/build_runner.zig
Error: /Users/dave/code/bun/build.zig:26:9: error: Bun requires Zig version 0.13.0.
/Users/dave/code/bun/build.zig:156:7: error: no field named 'zig_lib_dir' in struct 'Build'
/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib/std/Build.zig:1:1: note: struct declared here

error (store ): Failed to load build configuration for file:///Users/dave/code/bun/build.zig (error: error.RunFailed)
debug (server): Took 24ms to process notification-textDocument/didChange on Thread 2620455
debug (store ): Opened document 'file:///Users/dave/code/bun/root.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/root.zig'
debug (store ): Opened document 'file:///Users/dave/code/bun/root.zig'
debug (server): Took 18ms to process request-3-textDocument/documentSymbol on Thread 2620459
debug (store ): Resolved build file of 'file:///Users/dave/code/bun/src/bun.zig' as 'file:///Users/dave/code/bun/build.zig'
thread 2620459 panic: reached unreachable code
/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib/std/debug.zig:401:14: 0x10099cb47 in assert (zls)
    if (!ok) unreachable; // assertion failure
             ^
/Users/dave/code/zls/src/analysis.zig:602:21: 0x100b35fd7 in identifierLocFromIndex (zls)
    std.debug.assert(source_index < tree.source.len);
                    ^
/Users/dave/code/zls/src/features/references.zig:446:53: 0x100b3e2ff in referencesHandler (zls)
    const name_loc = Analyser.identifierLocFromIndex(handle.tree, source_index) orelse return null;
                                                    ^
/Users/dave/code/zls/src/Server.zig:1588:54: 0x100abdf2b in documentHighlightHandler (zls)
    const response = try references.referencesHandler(server, arena, .{ .highlight = request });
                                                     ^
/Users/dave/code/zls/src/Server.zig:1890:82: 0x100a4b01f in sendRequestSync__anon_22554 (zls)
        .@"textDocument/documentHighlight" => try server.documentHighlightHandler(arena, params),
                                                                                 ^
/Users/dave/code/zls/src/Server.zig:1953:58: 0x1009e3ce7 in processMessage (zls)
                const result = try server.sendRequestSync(arena_allocator.allocator(), @tagName(method), params);
                                                         ^
/Users/dave/code/zls/src/Server.zig:1967:33: 0x1009bb813 in processMessageReportError (zls)
    return server.processMessage(message) catch |err| {
                                ^
/Users/dave/code/zls/src/Server.zig:2005:62: 0x1009a13e3 in processJob (zls)
            const response = server.processMessageReportError(parsed_message.value) orelse return;
                                                             ^
/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib/std/Thread/Pool.zig:229:39: 0x1009a00d3 in runFn (zls)
            @call(.auto, func, closure.arguments);
                                      ^
/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib/std/Thread/Pool.zig:291:32: 0x100a33963 in worker (zls)
            run_node.data.runFn(&run_node.data, id);
                               ^
/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib/std/Thread.zig:409:13: 0x1009d6e6f in callFn__anon_13028 (zls)
            @call(.auto, f, args);
            ^
/Users/dave/zig/0.14.0-dev.1511+54b668f8a/files/lib/std/Thread.zig:678:30: 0x1009ad01b in entryFn (zls)
                return callFn(f, args_ptr.*);
                             ^
???:?:?: 0x183ab6f93 in ??? (libsystem_pthread.dylib)
???:?:?: 0xbb73000183ab1d33 in ??? (???)
```

--- a/src/DocumentScope.zig
+++ b/src/DocumentScope.zig
@@ -78,6 +78,7 @@ pub const DeclarationLookupContext = struct {
 };
 
 /// Assumes that the `node` is not a container_field of a struct tuple field.
+/// Returns a `.identifier` or `.builtin` token.
 fn getDeclNameToken(tree: Ast, node: Ast.Node.Index) ?Ast.TokenIndex {
     const tags = tree.nodes.items(.tag);
     const token_tags = tree.tokens.items(.tag);
@@ -130,8 +131,10 @@ fn getDeclNameToken(tree: Ast, node: Ast.Node.Index) ?Ast.TokenIndex {
     };
 
     if (token_index >= tree.tokens.len) return null;
-    if (token_tags[token_index] != .identifier) return null;
-    return token_index;
+    return switch (token_tags[token_index]) {
+        .identifier, .builtin => token_index,
+        else => null,
+    };
 }
 
 pub const Declaration = union(enum) {
@@ -254,6 +257,7 @@ pub const Declaration = union(enum) {
         return std.meta.eql(a, b);
     }
 
+    /// Returns a `.identifier` or `.builtin` token.
     pub fn nameToken(decl: Declaration, tree: Ast) Ast.TokenIndex {
         return switch (decl) {
             .ast_node => |n| getDeclNameToken(tree, n).?,
@@ -863,6 +867,7 @@ noinline fn walkContainerDecl(
                 container_field.convertToNonTupleLike(tree.nodes);
                 if (container_field.ast.tuple_like) continue;
                 const main_token = container_field.ast.main_token;
+                if (token_tags[main_token] != .identifier) continue;
                 try scope.pushDeclaration(main_token, .{ .ast_node = decl }, .field);
 
                 if (is_enum_or_tagged_union) {

--- a/src/analysis.zig
+++ b/src/analysis.zig
@@ -2073,8 +2073,7 @@ fn resolveTypeOfNodeUncached(analyser: *Analyser, node_handle: NodeWithHandle) e
 
             for (start..end + 1, 0..) |token_index, i| {
                 const slice = tree.tokenSlice(@intCast(token_index));
-                const carriage_return_ending: usize = if (slice[slice.len - 2] == '\r') 2 else 1;
-                length += slice.len - carriage_return_ending - 2 + @intFromBool(i != 0);
+                length += slice.len - 2 + @intFromBool(i != 0);
             }
 
             const string_literal_type = try analyser.ip.get(analyser.gpa, .{ .pointer_type = .{

--- a/src/analysis.zig
+++ b/src/analysis.zig
@@ -599,7 +599,7 @@ pub fn isSnakeCase(name: []const u8) bool {
 /// if the `source_index` points to `@name`, the source location of `name` without the `@` is returned.
 /// if the `source_index` points to `@"name"`, the source location of `name` is returned.
 pub fn identifierLocFromIndex(tree: Ast, source_index: usize) ?offsets.Loc {
-    std.debug.assert(source_index < tree.source.len);
+    std.debug.assert(source_index <= tree.source.len);
 
     var start = source_index;
     while (start > 0 and isSymbolChar(tree.source[start - 1])) {
@@ -617,17 +617,17 @@ pub fn identifierLocFromIndex(tree: Ast, source_index: usize) ?offsets.Loc {
 
 test identifierLocFromIndex {
     var tree = try Ast.parse(std.testing.allocator,
-        \\;name;  ;@builtin; ;@"escaped";
+        \\;name;  ;@builtin; ;@"escaped";end
     , .zig);
     defer tree.deinit(std.testing.allocator);
 
     try std.testing.expectEqualSlices(
         std.zig.Token.Tag,
         &.{
-            .semicolon, .identifier, .semicolon,
-            .semicolon, .builtin,    .semicolon,
-            .semicolon, .identifier, .semicolon,
-            .eof,
+            .semicolon,  .identifier, .semicolon,
+            .semicolon,  .builtin,    .semicolon,
+            .semicolon,  .identifier, .semicolon,
+            .identifier, .eof,
         },
         tree.tokens.items(.tag),
     );
@@ -649,6 +649,13 @@ test identifierLocFromIndex {
     try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 22, .end = 29 }), identifierLocFromIndex(tree, 22));
     try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 22, .end = 29 }), identifierLocFromIndex(tree, 25));
     try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 22, .end = 29 }), identifierLocFromIndex(tree, 29));
+
+    std.debug.assert(std.mem.eql(u8, "end", offsets.locToSlice(tree.source, .{ .start = 31, .end = 34 })));
+    try std.testing.expectEqual(@as(?offsets.Loc, null), identifierLocFromIndex(tree, 30));
+    try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 31, .end = 34 }), identifierLocFromIndex(tree, 31));
+    try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 31, .end = 34 }), identifierLocFromIndex(tree, 32));
+    try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 31, .end = 34 }), identifierLocFromIndex(tree, 33));
+    try std.testing.expectEqual(@as(?offsets.Loc, .{ .start = 31, .end = 34 }), identifierLocFromIndex(tree, 34));
 }
 
 /// Resolves variable declarations consisting of chains of imports and field accesses of containers

--- a/src/analysis.zig
+++ b/src/analysis.zig
@@ -3604,6 +3604,7 @@ pub const DeclWithHandle = struct {
         return a.decl.eql(b.decl) and std.mem.eql(u8, a.handle.uri, b.handle.uri);
     }
 
+    /// Returns a `.identifier` or `.builtin` token.
     pub fn nameToken(self: DeclWithHandle) Ast.TokenIndex {
         return self.decl.nameToken(self.handle.tree);
     }

--- a/src/offsets.zig
+++ b/src/offsets.zig
@@ -233,19 +233,24 @@ fn identifierIndexToLoc(tree: Ast, source_index: usize) Loc {
     return .{ .start = source_index, .end = index };
 }
 
+/// Support formats:
+/// - `foo`
+/// - `@"foo"`
+/// - `@foo`
 pub fn identifierIndexToNameLoc(text: [:0]const u8, source_index: usize) Loc {
-    if (text[source_index] == '@') {
-        std.debug.assert(text[source_index + 1] == '\"');
+    if (text[source_index] == '@' and text[source_index + 1] == '\"') {
         const start_index = source_index + 2;
         var index: usize = start_index;
         while (true) : (index += 1) {
-            if (text[index] == '\"') {
-                break;
+            switch (text[index]) {
+                '\n', '\"' => break,
+                else => {},
             }
         }
         return .{ .start = start_index, .end = index };
     } else {
         var index: usize = source_index;
+        if (text[index] == '@') index += 1;
         while (true) : (index += 1) {
             switch (text[index]) {
                 'a'...'z', 'A'...'Z', '_', '0'...'9' => {},
@@ -268,6 +273,8 @@ test identifierIndexToNameLoc {
     try std.testing.expectEqualStrings("hello", identifierIndexToNameSlice("@\"hello\"", 0));
     try std.testing.expectEqualStrings("hello", identifierIndexToNameSlice("@\"hello\" world", 0));
     try std.testing.expectEqualStrings("world", identifierIndexToNameSlice("@\"hello\" @\"world\"", 9));
+
+    try std.testing.expectEqualStrings("@hello", identifierIndexToNameSlice("@hello", 0));
 }
 
 pub fn identifierIndexToNameSlice(text: [:0]const u8, source_index: usize) []const u8 {
@@ -275,7 +282,11 @@ pub fn identifierIndexToNameSlice(text: [:0]const u8, source_index: usize) []con
 }
 
 pub fn identifierTokenToNameLoc(tree: Ast, identifier_token: Ast.TokenIndex) Loc {
-    std.debug.assert(tree.tokens.items(.tag)[identifier_token] == .identifier);
+    std.debug.assert(switch (tree.tokens.items(.tag)[identifier_token]) {
+        .builtin => true, // The Zig parser likes to emit .builtin where a identifier would be expected
+        .identifier => true,
+        else => false,
+    });
     return identifierIndexToNameLoc(tree.source, tree.tokens.items(.start)[identifier_token]);
 }
 

--- a/tests/lsp_features/folding_range.zig
+++ b/tests/lsp_features/folding_range.zig
@@ -278,7 +278,7 @@ test "multi-line string literal" {
         \\    \\world
         \\;
     , &.{
-        .{ .startLine = 1, .startCharacter = 4, .endLine = 3, .endCharacter = 0 },
+        .{ .startLine = 1, .startCharacter = 4, .endLine = 2, .endCharacter = 11 },
     });
 }
 

--- a/tests/lsp_features/hover.zig
+++ b/tests/lsp_features/hover.zig
@@ -224,7 +224,6 @@ test "string literal" {
         \\```zig
         \\const foo =
         \\    \\ipsum lorem
-        \\
         \\```
         \\```zig
         \\(*const [11:0]u8)
@@ -240,7 +239,6 @@ test "string literal" {
         \\const foo =
         \\    \\ipsum lorem
         \\    \\dolor sit amet
-        \\
         \\```
         \\```zig
         \\(*const [26:0]u8)

--- a/tests/lsp_features/semantic_tokens.zig
+++ b/tests/lsp_features/semantic_tokens.zig
@@ -1788,6 +1788,19 @@ test "weird code" {
         .{ "=", .operator, .{} },
         .{ "5", .number, .{} },
     });
+    try testSemanticTokens(
+        \\const foo = enum {
+        \\    @"a",
+        \\    @b,
+        \\};
+    , &.{
+        .{ "const", .keyword, .{} },
+        .{ "foo", .@"enum", .{ .declaration = true } },
+        .{ "=", .operator, .{} },
+        .{ "enum", .keyword, .{} },
+        .{ "@\"a\"", .enumMember, .{ .declaration = true } },
+        .{ "@b", .enumMember, .{ .declaration = true } },
+    });
 }
 
 const TokenData = struct {

--- a/tests/lsp_features/semantic_tokens.zig
+++ b/tests/lsp_features/semantic_tokens.zig
@@ -91,10 +91,9 @@ test "string literals" {
         .{ "const", .keyword, .{} },
         .{ "gamma", .variable, .{ .declaration = true } },
         .{ "=", .operator, .{} },
-        // TODO remove the newline
-        .{ "\\\\hello\n", .string, .{} },
-        .{ "\\\\world\n", .string, .{} },
-        .{ "\\\\\n", .string, .{} },
+        .{ "\\\\hello", .string, .{} },
+        .{ "\\\\world", .string, .{} },
+        .{ "\\\\", .string, .{} },
     });
 }
 

